plutus-scb/src/Cardano/Protocol/Socket/Server.hs

```diff
@@ -0,0 +1,152 @@
+{-# LANGUAGE DataKinds         #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs             #-}
+{-# LANGUAGE TypeFamilies      #-}
+{-# LANGUAGE TypeOperators     #-}
+module Cardano.Protocol.Socket.Server where
+
+import qualified Data.ByteString.Lazy                                as LBS
+import           Data.Functor.Contravariant                          (contramap)
+import           Data.List                                           (intersect, (!!))
+import           Data.Maybe                                          (listToMaybe)
+import           Data.Void                                           (Void)
+
+import           Control.Concurrent
+import           Control.Concurrent.Async
+import           Control.Lens
+import           Control.Tracer
+
+import qualified Ouroboros.Network.Protocol.ChainSync.Server         as ChainSync
+import qualified Ouroboros.Network.Protocol.LocalTxSubmission.Server as TxSubmission
+
+import           Network.Mux.Types                                   (AppType (..))
+import           Ouroboros.Network.Block                             (Point (..))
+import           Ouroboros.Network.Magic
+import           Ouroboros.Network.Mux
+import           Ouroboros.Network.NodeToNode
+import           Ouroboros.Network.Protocol.Handshake.Type
+import           Ouroboros.Network.Protocol.Handshake.Version
+import           Ouroboros.Network.Socket
+
+import           Codec.Serialise                                     (DeserialiseFailure)
+import           Cardano.Protocol.Socket.Type
+import           Cardano.Protocol
+import           Ledger                                              (Block, Tx (..))
+import           Wallet.Emulator.Chain
+
+startServerNode :: FilePath -> MVar ChainState -> IO Void
+startServerNode sockAddr stateVar = do
```

Comment:
I have no doubt that it took you quite some time to work out the right series of mystery incantations to put here - please document it so we don't forget why!